### ISSUE TITLE
Question about test result

### ISSUE THREAD

**Issue by zhqia** [shnhrtkyk/FWNet2#4]
I have trained the models, and tested the data using the trained models. But the result is much lower than the result in paper, even the precision and recall of transmission tower is zero. Besides, I find that when I set density 15 which is default in infer.py and knn 50000 which is written in paper, it can not contain all test data.
I wonder if there is some other difference in my code which cause the lower precision, or I should change the value of density and knn? And how should I set the value of density, is there some relationship between the value of density and knn?
In addition ,I find that the model in paper  do not concate the final feature in local and global part, but the model in the code is like that, I wonder which model I should use to reproduce the result of the paper.

**Comment by shnhrtkyk** [comment]
1. Imbalanced Data
In "train.py", I selected patches including minor classes like this.
        # if float(linespl[4]) > 0 and float(linespl[5]) > 0 and float(linespl[6]) > 0 and float(linespl[7]) > 0:

2. Density
If you chose higher density, you can predict class with overwrapped area.

3. model
The output of the model is combined, local and global modules. Please choose what you want.

**Comment by zhqia** [comment]
> 

Thank you for your answer. And another question about paper is that the conv1d in local part use 1×3 filter, however the filter in code is 160×1，I wonder which one I should use to achieve the result in paper.

**Comment by zhqia** [comment]
And I wonder what ' seg_pred_local, seg_pred_global, seg_point, seg_pointwave' represent in train.py in line 228?

**Comment by shnhrtkyk** [comment]
- About 2D Conv
The implementation at the time of posting is the tensorflow1.XX code for an old computer I was experimenting with.
I'm sharing the code that I rewrote it for Pytorch so that it works on modern computers.
It seems that I forgot to port conv2d when I rewrote the code.
At that time, I wrote tensor flow as follows.
`tf.nn.conv2d`

**Comment by shnhrtkyk** [comment]
- About `seg_pred_local, seg_pred_global, seg_point, seg_pointwave`
This is a remnant of experiments conducted under various conditions, so you can ignore it.

**Comment by zhqia** [comment]
Besides, I want to ask about loss function, it seems that the loss function in code is not as same as the description in paper. So how should I define the loss function?

**Comment by shnhrtkyk** [comment]
nll_loss(log_softmax(x), y) is the same as cross_entropy(softmax(x), y)
